content/post/deduplicating_pga_with_apollo.md

```diff
@@ -0,0 +1,387 @@
+
+The latter two features are not explicitly defined since they depend on several
+hyperparameters: maximum length of a feature, maximum number of random walks, etc.
+We used the maximum length to 3 and 4 for tree traversals and
+to 2 and 3 for random walks (the output is combined).
+The maximum number of random walks was 2 and the number of random steps was 10.
+Those values hold a trade-off between the vocabulary size and the representativeness.
+
+## Houston, we have a problem
+
+Sadly, it turned out that extracting those features crashed our cluster many, *many* 
+times. We faced several issues:
+
+- Running Spark on [Siva files](https://github.com/src-d/go-siva)
+through [jgit-spark-connector](https://github.com/src-d/jgit-spark-connector)
+led to massive amount of temporary data during feature extraction on the worker 
+pods of Kubernetes, which were further killed without notice by the cluster manager,
+- While converting the data to [Parquet](http://parquet.apache.org/) format to avoid this, 
+the workload imbalance due to [funny repos](https://github.com/Katee/git-bomb) made
+individual tasks last insanely long.
+- Once the conversion was done, applying TF-IDF to 1TB of cached data on disk
+led Spark to spend more time on garbage collection then on the actual work.
+- Finally, after we found a way to apply TF-IDF with manual batching (sic!),
+Babelfish failed to parse a considerable amount of the files.
+
+And that does not even take into account the continuous refactoring and bug detection
+in `src-d/ml`, `src-d/jgit-spark-connector`, `src-d/apollo` that any young project
+faces. We had to adapt rapidly to changes upstream.
+
+Nevertheless, after months of big data war, we managed to overcome the challenges
+and extract 6,194,874 distinct features from 7,869,917 million files, out of 102,114
+projects. The sparsity of that huge matrix was 0.00017, the average number
+of features per file was 1102. Not surprisingly, more than half of the processed files 
+consisted of JavaScript and Java, roughly the same 
+proportion as in the whole PGA. Overall, we were able to parse 45% to 67%
+of files in each language.
+
+|                       |  Python   |    Java   | Javascript |  Ruby   |    Go     |
+|-----------------------|-----------|-----------|------------|---------|-----------|
+| Files in PGA          | 1,673,172 | 4,021,258 | 5,549,381  | 909,935 | 1,962,502 |
+| % of all PGA          | 2.47      | 5.93      | 8.18       | 1.34    | 2.89      |
+| Processed             | 1,021,687 | 1,848,084 | 3,066,721  | 628,081 | 1,305,344 |
+| Ratio processed/PGA, %| 61.06     | 45.95     | 55.26      | 69.02   | 66.51     |
+
+{{% caption src="/post/deduplicating_pga_with_apollo/languages.png" %}}
+Percent of files in each language in our corpus.
+{{% /caption %}}
+
+We observed that over 65% of distinct features were **literals** due high variance,
+and apparently were relevant enough for most of the files to be retained. Similarly, 
```

Comment:
It would be good to specify a little bit how tf-idf was used in this context (how it lead to non relevant features being discarded) because it was mentioned before but once and quite a while ago at this point in the blog post so this might come out of the blue for a reader.

Reply:
not sure how to do that w/o being too verbose and just kinda repeating the wiki link. personally I think it's pretty basic knowledge, or at least not hard to grasp how it was done. also we speak about it not so long ago, like 30 2 paragraphs above

Reply:
If there's some interesting note on this part of the process I'd definitely add a link to the other resource. Even without much context: "To learn more about how TF-IDF led to discarding non relevant features read [blah blah blah](link)"